tokio/src/net/addr.rs

```diff
@@ -40,6 +40,160 @@ where
+    pub struct LookupHost<T>
+    where
+        T: sealed::ToSocketAddrsPriv,
+        T: ToSocketAddrs<Future = sealed::MaybeReady>,
+        T: ToSocketAddrs<Iter = sealed::OneOrMore>,
+    {
+        inner: LookupHostInner<T>,
+    }
+
+    impl<T> Future for LookupHost<T>
+    where
+        T: sealed::ToSocketAddrsPriv,
+        T: ToSocketAddrs<Future = sealed::MaybeReady>,
+        T: ToSocketAddrs<Iter = sealed::OneOrMore>,
+    {
+        type Output = io::Result<SocketAddr>;
+
+        fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            if let LookupHostInner::Pending { fut } = &mut self.inner {
+                if let Poll::Ready(addrs) = Pin::new(fut).poll(cx) {
+                    match addrs {
+                        Ok(mut addrs) => {
+                            let addr = addrs.next().unwrap();
+                            Poll::Ready(Ok(addr))
+                        },
+                        Err(e) => Poll::Ready(Err(e))
+                    }
+                } else {
+                    Poll::Pending
+                }
+            } else {
+                unimplemented!()
+            }
+        }
+    }
+
+    // TODO: collapse this into a single enum once Rust 1.40 ships
+    // with the the `#[non_exhaustive]` attribute if Tokio's MSRV allows
+    //
+    // https://github.com/rust-lang/rust/pull/64639
+    #[derive(Debug)]
+    enum LookupHostInner<T>
+    where
+        T: sealed::ToSocketAddrsPriv,
+        T: ToSocketAddrs<Future = sealed::MaybeReady>,
+        T: ToSocketAddrs<Iter = sealed::OneOrMore>,
+    {
+        Pending {
+            fut: <T as sealed::ToSocketAddrsPriv>::Future,
+        },
+        Addrs {
+            addrs: sealed::OneOrMore,
+        },
+    }
+
+    impl<T> LookupHost<T>
+    where
+        T: sealed::ToSocketAddrsPriv + Unpin,
+        T: ToSocketAddrs<Future = sealed::MaybeReady>,
+        T: ToSocketAddrs<Iter = sealed::OneOrMore>,
+    {
+        /// Fetches the next resolved address. If the first call to `next_addr` doesn't
+        /// return an error, then every subsequent call to `next_addr` will succeed.
+        /// Once this method returns `Ok(None)`, `LookupHost` should be considered exhausted
+        /// as all `SocketAddr`s have been yielded.
+        pub async fn next_addr(&mut self) -> io::Result<Option<SocketAddr>> {
+            let inner = &mut self.inner;
+            if let LookupHostInner::Pending { fut } = inner {
+                *inner = LookupHostInner::Addrs { addrs: fut.await? };
+            }
+
+            match inner {
+                LookupHostInner::Addrs { addrs } => Ok(addrs.next()),
+                _ => unreachable!("The future should have already been resolved"),
+            }
+        }
+    }
+
+    /// Performs a DNS resolution.
+    ///
+    /// This API is not intended to cover all DNS use cases.
+    /// Anything beyond the basic use case should be done with a specialized library.
```

Comment:
Based on this statement, I'm not sure why this needs to have such a "complicated" API surface.

### One **understanding** from my side (correct me if wrong):

All mainstream DNS resolvers return all resolved addresses synchronously. Therefore just returning `Vec<SocketAddr>` or `Vec<LookupResult>` (if we need more information) should be fine.

### One **experience report** from my side:

Most applications only care about 1 DNS result. However in the rare situations where I wanted more I always wanted to have all of them synchronously, so that I can filter the list and decide for the best result. An async iterator wouldn't really have bought me a lot here.

Based on these observations I think just providing `async fn lookup_host(host: &str) -> Vec<LookupResult>` as a default is fine.

If you want application writers to have more control and e.g. to avoid allocations, you could provide an optional `LookupOptions` struct to the function which e.g. allows to configure a predicate for filtering results before they are returned to the user. E.g.

```rust 
struct LookupOptions {
    /// max amount of results to return
    limit: usize,
    /// filters results before returned
    filter: Fn(&LookupResult) -> bool,
}

async fn lookup_host_ext(host: &str, options: LookupOptions) -> Vec<LookupResult>
/// Not a great name, but you get the idea
async fn lookup_host_first_ext(host: &str, options: LookupOptions) -> LookupResult
```

Reply:
@Matthias247 one argument I heard for returning multiple is that, some applications, want to get the IPv6 address and have to iterate through the results.

Reply:
Sure, that's the filtering use-case I mentioned. But given the state of DNS resolvers I think that doing the synchronous lookup and then a `.filter` should be fine for those use-cases. And if there gets support integrated for DNS resolvers which operate asynchronously (merge results from multiple sources, where on is available earliers), then adding the second API that I mentioned which applies filtering internally also seems fine.